### PR TITLE
FileUpload addon Blog

### PR DESCRIPTION
**Pull request by preetkaran20**
Signed-off-by: Karan Preet Singh Sasan <preetkaran20@gmail.com>

**- Summary**
[Blog on FileUpload Addon](https://github.com/SasanLabs/owasp-zap-fileupload-addon)

As we are releasing FileUpload addon so this blog explains the addon configurations
**- Test plan**

image

**- Description for the changelog**
Blog post on FileUpload Addon

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: site/content/blog/2021-08-16-zap-fileupload-addon/index.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+type: post
+tags:
+- blog
+date: "2021-08-16"
+addSocialPreview: true
+authors:
+    - preetkaran20
+---
+## Overview
+File upload is becoming a more and more essential part of any application, where the user is able to upload their photo, their CV, or a video showcasing a project they are working on. The application should be able to fend off bogus and malicious files in a way to keep the application and the users safe. Generally file upload functionality is quite complex to automate and has huge attack surface hence there is a need to automate the process and also secure it. So the FileUpload add-on has scan rule which is used to find vulnerabilities in File upload functionality and this blog explains on how to use it.
```
Comment (thc202) [inline site/content/blog/2021-08-16-zap-fileupload-addon/index.md:15]: The add-on was actually named `File upload Scan Rule`, either change the name here or in the add-on.

File: site/content/blog/2021-08-16-zap-fileupload-addon/index.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+- JPEG and GIF images
+- EICAR file
+- SVG images
+
+Vulnerabilities covered are:
+- Cross site scripting
+- Remote Code Execution by uploading JSP and its variants
+- No virus scan. 
+
+The FileUpload add-on is available in the marketplace and can be installed from within ZAP.
```
Comment (thc202) [inline site/content/blog/2021-08-16-zap-fileupload-addon/index.md:49]: It should be more clear that's only available for dev versions (weekly and built from source). The add-on was not yet released though.

File: site/content/blog/2021-08-16-zap-fileupload-addon/index.md (hunk 1 above)
Comment (preetkaran20) [reply site/content/blog/2021-08-16-zap-fileupload-addon/index.md:15]: Done

File: site/content/blog/2021-08-16-zap-fileupload-addon/index.md (hunk 2 above)
Comment (preetkaran20) [reply site/content/blog/2021-08-16-zap-fileupload-addon/index.md]: Done.

File: site/content/blog/2021-08-16-zap-fileupload-addon/index.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+3. In some applications the URL to retrieve the uploaded file is present in the **response of a different URL** which is called a **preflight request**. E.g. Profile picture URL is part of profile page and hence we need to parse the response of the profile page to find the URL of the profile picture. For handling this type of configuration, the options panel has **Dynamic Location Configuration** which has a **URI Regex** and **Parse Http Response Configuration** which has **Start Identifier**, and **End Identifier**. So the File Upload add-on will invoke the URI mentioned in URI Regex and then parse the response using Start Identifier and End Identifier. URI Regex field also supports the dynamic file name by **${fileName}**
+
+As FileUpload add-on's scan rule will execute as part of Owasp ZAP Active Scan, so after filling these configuration details, run the active scan in order to find the vulnerabilities in FileUpload Functionality.
```
Comment (psiinon) [inline site/content/blog/2021-08-16-zap-fileupload-addon/index.md:32]: Owasp -> OWASP

**Comment by psiinon** [comment]
@preetkaran20 do you have / know of a vulnerable app that people could use to test this on? 
If so then I think a walk through of what the user needs to do would be really helpful.
You could also explain how this could be completely automated, ie using the ZAP `-config` option to set the required URLs. I cant see anyone using this in automation unless they know how to do that.

Also ... you can add a photo of yourself if you like :)
That would go under here: https://github.com/zaproxy/zaproxy-website/tree/master/site/static/img/authors

File: site/content/blog/2021-08-16-zap-fileupload-addon/index.md (hunk 3 above)
Comment (preetkaran20) [reply site/content/blog/2021-08-16-zap-fileupload-addon/index.md:32]: Sure

**Comment by preetkaran20** [comment]
> @preetkaran20 do you have / know of a vulnerable app that people could use to test this on?
> If so then I think a walk through of what the user needs to do would be really helpful.
> You could also explain how this could be completely automated, ie using the ZAP `-config` option to set the required URLs. I cant see anyone using this in automation unless they know how to do that.
> 
> Also ... you can add a photo of yourself if you like :)
> That would go under here: https://github.com/zaproxy/zaproxy-website/tree/master/site/static/img/authors

We can test on https://github.com/SasanLabs/VulnerableApp-facade. Shall i add a blog on that? or Shall i add a video? I will look at automation thing. Sure will update the photo.

File: site/content/blog/2021-08-16-zap-fileupload-addon/index.md (hunk 3 above)
Comment (preetkaran20) [reply site/content/blog/2021-08-16-zap-fileupload-addon/index.md]: Done

**Comment by preetkaran20** [comment]
> > @preetkaran20 do you have / know of a vulnerable app that people could use to test this on?
> > If so then I think a walk through of what the user needs to do would be really helpful.
> > You could also explain how this could be completely automated, ie using the ZAP `-config` option to set the required URLs. I cant see anyone using this in automation unless they know how to do that.
> > Also ... you can add a photo of yourself if you like :)
> > That would go under here: https://github.com/zaproxy/zaproxy-website/tree/master/site/static/img/authors
> 
> We can test on https://github.com/SasanLabs/VulnerableApp-facade. Shall i add a blog on that or details it here only? or Shall i add a video? I will look at automation thing. Sure will update the photo.

**Comment by psiinon** [comment]
> > @preetkaran20 do you have / know of a vulnerable app that people could use to test this on?
> > If so then I think a walk through of what the user needs to do would be really helpful.
> > You could also explain how this could be completely automated, ie using the ZAP `-config` option to set the required URLs. I cant see anyone using this in automation unless they know how to do that.
> > Also ... you can add a photo of yourself if you like :)
> > That would go under here: https://github.com/zaproxy/zaproxy-website/tree/master/site/static/img/authors
> 
> We can test on https://github.com/SasanLabs/VulnerableApp-facade. Shall i add a blog on that or detail it here only? or Shall i add a video? I will look at automation thing. Sure will update the photo.

Entirely up to you. But the easier it looks the more likely people will try it out :)

**Comment by preetkaran20** [comment]
> > > @preetkaran20 do you have / know of a vulnerable app that people could use to test this on?
> > > If so then I think a walk through of what the user needs to do would be really helpful.
> > > You could also explain how this could be completely automated, ie using the ZAP `-config` option to set the required URLs. I cant see anyone using this in automation unless they know how to do that.
> > > Also ... you can add a photo of yourself if you like :)
> > > That would go under here: https://github.com/zaproxy/zaproxy-website/tree/master/site/static/img/authors
> > 
> > 
> > We can test on https://github.com/SasanLabs/VulnerableApp-facade. Shall i add a blog on that or detail it here only? or Shall i add a video? I will look at automation thing. Sure will update the photo.
> 
> Entirely up to you. But the easier it looks the more likely people will try it out :)

Then will create another blog explaining, how to use the addon, using it in Automation scripts, and also might add a blog for JWT addon too.

**Comment by kingthorin** [comment]
I haven't really read through this yet, but based on messages/comments let's make sure it's clear that this will work with recent and future weekly builds (not just dev).

**Review by kingthorin — COMMENTED** [review]
Minor stuff

File: site/content/blog/2021-08-16-zap-fileupload-addon/index.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+3. In some applications the URL to retrieve the uploaded file is present in the **response of a different URL** which is called a **preflight request**. E.g. Profile picture URL is part of profile page and hence we need to parse the response of the profile page to find the URL of the profile picture. For handling this type of configuration, the options panel has **Dynamic Location Configuration** which has a **URI Regex** and **Parse Http Response Configuration** which has **Start Identifier**, and **End Identifier**. So the File Upload add-on will invoke the URI mentioned in URI Regex and then parse the response using Start Identifier and End Identifier. URI Regex field also supports the dynamic file name by **${fileName}**
+
+As FileUpload add-on's scan rule will execute as part of OWASP ZAP Active Scan, so after filling these configuration details, run the active scan in order to find the vulnerabilities in FileUpload Functionality.
```
[the remark this answers is not in the document]
Comment (preetkaran20) [reply site/content/blog/2021-08-16-zap-fileupload-addon/index.md:32]: yeah good point. I think i have not added the check. I will add and raise a PR.

[the remark this answers is not in the document]
Comment (preetkaran20) [reply site/content/blog/2021-08-16-zap-fileupload-addon/index.md:32]: Raised a PR: https://github.com/SasanLabs/owasp-zap-fileupload-addon/pull/7

please review.

[the remark this answers is not in the document]
Comment (kingthorin) [reply site/content/blog/2021-08-16-zap-fileupload-addon/index.md:32]: Thanks.

**Comment by preetkaran20** [comment]
Hi @psiinon,

I have create a video: https://www.youtube.com/watch?v=3bHjrpbLQmA which explains the FileUpload addon. Please have a look.

thanks,
Karan

**Comment by preetkaran20** [comment]
Somehow Video playback is not working in ZAP website: https://www.zaproxy.org/blog/2017-08-22-zap-browser-launch/
is it working for you guys?
image

**Comment by psiinon** [comment]
> Somehow Video playback is not working in ZAP website: https://www.zaproxy.org/blog/2017-08-22-zap-browser-launch/
> is it working for you guys?

Oops, thats a bug. A couple of places we're using `{{< youtube lYHwR7UEggA >}}` (old style) where we shoiuld be using `{{< youtube uuid="lYHwR7UEggA" >}}` - I'll submit a PR tomorrow...

**Comment by preetkaran20** [comment]
Hi @psiinon,

I have correct it. Please review the PR.

thanks,
Karan

**Comment by psiinon** [comment]
Looks good, and I particularly like your video - I think its really clear and helpful!
You will need to change the date and you could add another tag like "fileupload".
Other than that it all looks good to me, and I'd be v happy for it to be released today.

**Comment by psiinon** [comment]
Oh, and I submitted another PR for the youtube link fixes: https://github.com/zaproxy/zaproxy-website/pull/522 - there was another one that needed changing as well ;)

File: site/content/blog/2021-08-16-zap-fileupload-addon/index.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+type: post
+tags:
+- blog
+date: "2021-08-16"
+addSocialPreview: true
+authors:
+    - preetkaran20
+---
+## Overview
+File upload is becoming a more and more essential part of any application, where the user is able to upload their photo, their CV, or a video showcasing a project they are working on. The application should be able to fend off bogus and malicious files in a way to keep the application and the users safe. Generally file upload functionality is quite complex to automate and has huge attack surface hence there is a need to automate the process and also secure it. So the FileUpload add-on has scan rule which is used to find vulnerabilities in File upload functionality and this blog explains on how to use it.
```
Comment (thc202) [inline site/content/blog/2021-08-16-zap-fileupload-addon/index.md:15]: `in File upload` lower case file.

File: site/content/blog/2021-08-16-zap-fileupload-addon/index.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+## Configuration
+File upload functionality generally has 2 endpoints, one from where file is uploaded and one from where file is retrieved. It is necessary to know both these endpoints. While Active Scanning an application, file upload endpoint is already known but retrieval endpoint is not known to the scan rule hence there are configuration details specific to the retrieval endpoint.
+
+Under ZAP's Options dialog you will find a FileUpload section as shown below:
```
Comment (thc202) [inline site/content/blog/2021-08-16-zap-fileupload-addon/index.md:20]: `the File Upload screen`

File: site/content/blog/2021-08-16-zap-fileupload-addon/index.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+2. In some applications the URL to retrieve the uploaded file is present in the **file upload request's response**. For handling this type of configuration, options panel has **Parse HTTP Response Configuration** which has 2 parameters **Start Identifier** and **End Identifier**. These identifiers are used to locate the URL within the response.
+3. In some applications the URL to retrieve the uploaded file is present in the **response of a different URL** which is called a **preflight request**. E.g. Profile picture URL is part of profile page and hence we need to parse the response of the profile page to find the URL of the profile picture. For handling this type of configuration, the options panel has **Dynamic Location Configuration** which has a **URI Regex** and **Parse HTTP Response Configuration** which has **Start Identifier**, and **End Identifier**. So the File Upload add-on will invoke the URI mentioned in URI Regex and then parse the response using Start Identifier and End Identifier. URI Regex field also supports the dynamic file name by **${fileName}**.
```
Comment (thc202) [inline site/content/blog/2021-08-16-zap-fileupload-addon/index.md:30]: `FileUpload add-on`

File: site/content/blog/2021-08-16-zap-fileupload-addon/index.md [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+As FileUpload add-on's scan rule will execute as part of OWASP ZAP Active Scan, so after filling these configuration details, run the active scan in order to find the vulnerabilities in FileUpload Functionality.
```
Comment (thc202) [inline site/content/blog/2021-08-16-zap-fileupload-addon/index.md:32]: `file upload functionality`

**Comment by thc202** [comment]
Note that you still need to rename the files for the newer date.

File: site/content/blog/2017-08-22-zap-browser-launch/index.md [hunk 9]
```diff
@@ -22,7 +22,7 @@ You will now get a new 'Launch Browser' option in the Quick Start tab:
   
 {{< img "images/launch-browser.png" >}}  
 To see a demo of this feature see the following video:
-{{< youtube lYHwR7UEggA >}}
+{{< youtube uuid="lYHwR7UEggA" >}}
```
Comment (thc202) [inline site/content/blog/2017-08-22-zap-browser-launch/index.md:25]: This can be reverted, being fixed in the other pull request.

**Comment by preetkaran20** [comment]
> Note that you still need to rename the directory for the newer date.

oh man .... too need to open eclipse again

File: site/content/blog/2017-08-22-zap-browser-launch/index.md (hunk 9 above)
Comment (preetkaran20) [reply site/content/blog/2017-08-22-zap-browser-launch/index.md]: done

File: site/content/blog/2021-08-20-zap-fileupload-addon/index.md [hunk 10]
```diff
@@ -0,0 +1,53 @@
+---
+title: "ZAP FileUpload Add-on"
+description: "How to use ZAP FileUpload Add-on for finding Vulnerabilities in file upload functionality"
+images:
+- https://www.zaproxy.org/blog/2021-08-16-zap-fileupload-addon/images/fileupload-options-panel.png
```
Comment (thc202) [inline site/content/blog/2021-08-20-zap-fileupload-addon/index.md:5]: Needs update for the date.

Comment (preetkaran20) [reply site/content/blog/2021-08-20-zap-fileupload-addon/index.md:5]: oh thanks for mentioning

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Review by psiinon — APPROVED** [review]

**Pull request merged**